lib/Runtime/Language/PrototypeChainCache.cpp

```diff
@@ -0,0 +1,361 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeLanguagePch.h"
+
+using namespace Js;
+
+template <typename T>
+ThreadCacheRegistry<T>::ThreadCacheRegistry(_In_ PageAllocator* pageAllocator) :
+    arena(_u("PrototypeChainCacheRegistry"), pageAllocator, Throw::OutOfMemory)
+{
+}
+
+template <typename T>
+ScriptCacheRegistry<T>**
+ThreadCacheRegistry<T>::Register(_In_ ScriptCacheRegistry<T>* cache)
+{
+    return this->scriptRegistrations.PrependNode(&this->arena, cache);
+}
+
+template <typename T>
+void
+ThreadCacheRegistry<T>::Unregister(_In_ ScriptCacheRegistry<T>** registration)
+{
+    this->scriptRegistrations.RemoveElement(&this->arena, registration);
+}
+
+template <typename T>
+void
+ThreadCacheRegistry<T>::Reset()
+{
+    this->scriptRegistrations.Reset();
+}
+
+template <typename T>
+void
+ThreadCacheRegistry<T>::Clear()
+{
+    bool hasItem = false;
+    FOREACH_DLISTBASE_ENTRY(ScriptCacheRegistry<T>*, registration, &this->scriptRegistrations)
+    {
+        registration->Clear(false /* unregister */);
+        hasItem = true;
+    }
+    NEXT_DLISTBASE_ENTRY;
+
+    if (!hasItem)
+    {
+        return;
+    }
+
+    this->scriptRegistrations.Reset();
+    this->arena.Reset();
+}
+
+template <typename T>
+ScriptCacheRegistry<T>::ScriptCacheRegistry(_In_ ThreadCacheRegistry<T>* threadRegistry) :
+    threadRegistry(threadRegistry),
+    cache(nullptr),
+    registration(nullptr)
+{
+}
+
+template <typename T>
+void
+AssignCache(_In_ T* cache)
+{
+    Assert(this->cache == nullptr);
+    this->cache = cache;
+}
+
+template <typename T>
+void
+ScriptCacheRegistry<T>::Register()
+{
+    Assert(this->cache != nullptr);
+    if (this->registration == nullptr)
+    {
+        this->registration = this->threadRegistry->Register(this);
+    }
+}
+
+template <typename T>
+void
+ScriptCacheRegistry<T>::Clear(bool unregister)
+{
+    if (this->registration != nullptr)
+    {
+        Assert(this->cache);
+        this->cache->Clear();
+        if (unregister)
+        {
+            this->threadRegistry->Unregister(this->registration);
+        }
+        this->registration = nullptr;
```

Comment:
I feel like it would be most consistent to only set registration=nullptr if we unregistered